### ISSUE TITLE
Encryption in transit: ssl, tls, SSL/TLS and STARTTLS

### ISSUE THREAD

**Issue by mendel5** [imapsync/imapsync#292]
Hi @gilleslamiral , thank you for creating and maintaining imapsync! It is an incredibely useful tool.

I have a question regarding encryption in transit when transferring e-mails. imapsync supports two parameters for encryption, these are:
- `--ssl1` / `--ssl2`
- `--tls1` / `--tls2`

I was not sure what these parameters were doing exactly because they might have a different meaning depending on the context, either in an encryption context or an email context. Let me cite a section from [https://github.com/imapsync/imapsync/blob/master/FAQ.d/FAQ.Security.txt](https://github.com/imapsync/imapsync/blob/master/FAQ.d/FAQ.Security.txt):

```
SSL: going to encryption before the imap session start. 
     It is on port 993.

TLS: going to encryption after the imap session start but before
     the credential are sent. It is on port 143.
```

Is the following description correct?
- The imapsync parameters `--ssl1` and `--ssl2` **force** the usage of the Secure Sockets Layer/ Transport Layer Security (SSL/TLS) protocol family. The protocol can be specified, for example by using: `--sslargs1 SSL_version=TLSv1_3`.
- The imapsync parameters `--tls1` and `--tls2` **force** the usage of the STARTTLS protocol (Opportunistic TLS).

According to [https://serverfault.com/questions/523804/is-starttls-less-safe-than-tls-ssl](https://serverfault.com/questions/523804/is-starttls-less-safe-than-tls-ssl), SSL/TLS is preferred over STARTTLS. Therefore imapsync's `--ssl` parameters are preferred over imapsync's `--tls` parameters.

If the descriptions above are correct, I would like to add these details to the Security FAQ at [https://github.com/imapsync/imapsync/blob/master/FAQ.d/FAQ.Security.txt](https://github.com/imapsync/imapsync/blob/master/FAQ.d/FAQ.Security.txt).

**Comment by gilleslamiral** [comment]
> I would like to add these details to the Security FAQ  FAQ.Security.txt

Please do!

**Comment by mendel5** [comment]
> Please do!

Thanks! But is everything I said correct? I want to make sure to only add correct information to the documentation.

**Comment by gilleslamiral** [comment]
Start from
https://imapsync.lamiral.info/FAQ.d/FAQ.Security.txt
as it is more recent.

For the correctness I'll read your patch, verify by myself and credit the passage to you so if you're not sure enough of what you write, don't write or better : write you're not sure of it. Notice that being sure of what we know is not always related to the truthiness of it :-)

**Comment by mendel5** [comment]
Ok, I will start working on this.

**Comment by mendel5** [comment]
It seems that this repository on Github ([https://github.com/imapsync/imapsync](https://github.com/imapsync/imapsync)) is not the main repository you use for development, right? Otherwise your website would not be on a newer version.

**Comment by gilleslamiral** [comment]
Latest imapsync is https://imapsync.lamiral.info/imapsync

**Comment by mendel5** [comment]
Ok, good to know. Then I will not open a pull request but simply paste my new content here in this issue.

**Comment by gilleslamiral** [comment]
As you wish, pull request is the normal way.

**Comment by mendel5** [comment]
All right, I will open a pull request.

**Comment by mendel5** [comment]
I opened a pull request at https://github.com/imapsync/imapsync/pull/294 , please have a look.

**Comment by gilleslamiral** [comment]
Thanks Mendel for your input!

> Parameters --tls1 and --tls2: Try to establish an encrypted connection based on STARTTLS. 

Yes and not. It's not "try to". It's "Establish an encrypted connection with STARTTLS."

> Does not fail if an encryption cannot be established 

Imapsync does fail if an encryption cannot be established:
````
sub login_imap
{
...
        if ( $tls ) {
                set_tls( $imap, $h ) ;
                if ( ! $imap->starttls(  ) )
                {
                        $mysync->{nb_errors}++ ;
                        exit_clean( $mysync, $EXIT_TLS_FAILURE,
                                "$Side failure: Can not go to tls encryption on $side [$host]:",
                                $imap->LastError, "\n"
                        ) ;
                }
                myprint( "$Side: Socket successfuly converted to SSL\n"  ) ;
        }
`````

> meaning an encrypted connection cannot be guaranteed [UNSURE]. 

encrypted guaranteed with imapsync with --tls1 and --tls2

> Less secure than --ssl1 and --ssl2.

Maybe, but not for that reason.

**Comment by mendel5** [comment]
> encrypted guaranteed with imapsync with --tls1 and --tls2

Ok, then imapsync's implementation seems to be more secure than the "normal" STARTTLS protocol (which is of course a good thing). For example the page at [https://en.wikipedia.org/wiki/Opportunistic_TLS#Weaknesses_and_mitigations](https://en.wikipedia.org/wiki/Opportunistic_TLS#Weaknesses_and_mitigations) says there's an attack against STARTTLS called `STRIPTLS attack`.

> Less secure than --ssl1 and --ssl2.
> Maybe, but not for that reason.

Do you mean that using the SSL parameters is more secure because they encrypt the connection before the IMAP session starts while the TLS parameters encrypt the connection after the IMAP session starts (as explained [here](https://github.com/imapsync/imapsync/blob/452ebe34ac4dca7f93c6cdee164dd9bf3873b9f4/FAQ.d/FAQ.Security.txt#L13))?

**Comment by gilleslamiral** [comment]
>  STRIPTLS attack might not be successful against imapsync's STARTTLS implementation because you said that Imapsync does fail if an encryption cannot be established.

Yes.

> Do you mean that using the SSL parameters is more secure because they encrypt the connection before the IMAP session starts while the TLS parameters encrypt the connection after the IMAP session starts

Yes. For example 

    imapsync --host1 5.39.87.81  --sslargs1 SSL_verify_mode=1 --ssl1 

fails with "fatal SSL error: hostname verification failed" while 

    imapsync --host1 5.39.87.81  --sslargs1 SSL_verify_mode=1 --tls1

succeed with "not using SNI because hostname is unknown" which is normal because
the resolution has been done long before, the hostname cannot be known as a name in STARTTLS context.

**Comment by mendel5** [comment]
Thanks for the explanation!

If you need anything else from me related to this issue or the pull request linked above, please let me know.

**Comment by gilleslamiral** [comment]
Regarding our discussion, I have updated
https://imapsync.lamiral.info/FAQ.d/FAQ.Security.txt
with your input.

I severally changed your text but I updated the whole document to minimize the possible confusion brought by --tls and --ssl and the SSL/TLS meaning. Have in mind that few people know about SSL and TLS distinction, only security IT focus people like you. 

So feel free to review it completely.
